pkg/cloudprovider/providers/vsphere/vsphere_util.go

```diff
@@ -63,3 +80,112 @@ func GetgovmomiClient(cfg *VSphereConfig) (*govmomi.Client, error) {
 	client, err := newClient(context.TODO(), cfg)
 	return client, err
 }
+
+// Get list of compatible datastores that satisfies the storage policy requirements.
+func (vs *VSphere) GetCompatibleDatastores(ctx context.Context, pbmClient *pbm.Client, resourcePool *object.ResourcePool, storagePolicyID string) ([]mo.Datastore, error) {
+	datastores, err := vs.getAllAccessibleDatastoresForK8sCluster(ctx, resourcePool)
+	if err != nil {
+		return nil, err
+	}
+	var hubs []pbmtypes.PbmPlacementHub
+	for _, ds := range datastores {
+		hubs = append(hubs, pbmtypes.PbmPlacementHub{
+			HubType: ds.Type,
+			HubId:   ds.Value,
+		})
+	}
+	req := []pbmtypes.BasePbmPlacementRequirement{
+		&pbmtypes.PbmPlacementCapabilityProfileRequirement{
+			ProfileId: pbmtypes.PbmProfileId{
+				UniqueId: storagePolicyID,
+			},
+		},
+	}
+	res, err := pbmClient.CheckRequirements(ctx, hubs, nil, req)
+	if err != nil {
+		return nil, err
+	}
+	compatibleHubs := res.CompatibleDatastores()
+	// Return an error if there are no compatible datastores.
+	if len(compatibleHubs) < 1 {
+		return nil, fmt.Errorf("There are no compatible datastores: %+v that satisfy the storage policy: %+q requirements", datastores, storagePolicyID)
```

Comment:
Returning 0 compatible datastores with nil error is a perfectly fine behavior for this method. The caller can decide if it was expecting 0 or 1 or 2 or more etc. Based on the caller's requirements, it can treat it as error or perfectly fine condition.